File: SW/apps/equalize/equalize.cpp
//----------------------------------------------------------------------------
// Copyright [2014] [Ztachip Technologies Inc]
//
// Author: Vuong Nguyen
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except IN compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to IN writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//------------------------------------------------------------------------------

#include <math.h>
#include "../../base/types.h"
#include "../../base/tensor.h"
#include "../../base/graph.h"
#include "kernels/equalize.h"
#include "equalize.h"


// Graph node to do image equalization (improve image contrast)
// Calculate pixel value histogram and then apply image
// equalization.

GraphNodeEqualize::GraphNodeEqualize() {
   m_spu=0;
}

GraphNodeEqualize::GraphNodeEqualize(TENSOR *input,TENSOR *output) : GraphNodeEqualize() {
   Create(input,output);
}

GraphNodeEqualize::~GraphNodeEqualize() {
   Cleanup();
}

ZtaStatus GraphNodeEqualize::Create(TENSOR *input,TENSOR *output) {
   Cleanup();
   m_input=input;
   m_output=output;
   return ZtaStatusOk;
}


ZtaStatus GraphNodeEqualize::Verify() {
   if((*(m_input->GetDimension())).size() != 3)
      return ZtaStatusFail;
   m_w=(*(m_input->GetDimension()))[2];
   m_h=(*(m_input->GetDimension()))[1];
   m_nChannel=(*(m_input->GetDimension()))[0];
   SetContrast(2.0);
   std::vector<int> dim={m_nChannel,2,kHistogramBinSize,VECTOR_WIDTH};
   m_result.Create(TensorDataTypeInt16,TensorFormatSplit,TensorSemanticUnknown,dim);
   std::vector<int> dim2={m_nChannel,m_h,m_w};
   m_output->Create(TensorDataTypeUint8,TensorFormatSplit,m_input->GetSemantic(),dim2);
   m_histogramAvail=false;
   return ZtaStatusOk;
}

ZtaStatus GraphNodeEqualize::Execute(int queue,bool stepMode) {
   if(m_histogramAvail) {
      GenEqualizer();
   }
   kernel_equalize_exe(
      (unsigned int)GetJobId(queue),
	  (unsigned int)m_input->GetBuf(),
	  (unsigned int)m_result.GetBuf(),
	  (unsigned int)m_output->GetBuf(),
      m_nChannel,
	  (unsigned int)((m_histogramAvail && m_contrast>0)?m_spu:0),
      m_w,
      m_h
   );
   m_histogramAvail=true;
   return ZtaStatusOk;
}

ZTA_SHARED_MEM GraphNodeEqualize::GenEqualizer() {
   int16_t *histogram;
   int count;
   bool flag;
   int overflow;

   histogram = (int16_t *)m_result.GetBuf();
   for (int i=0;i < HISTOGRAM_BIN_COUNT;i++) {
      m_histogram[i]=0;
      for(int j=0;j < m_nChannel;j++) {
         m_histogram[i] += HISTOGRAM(histogram+j*64,i);
      }
   }
   for (int j = 0;j < 8;j++) {
      flag = false;
      for (int i=0;i < HISTOGRAM_BIN_COUNT;i++) {
         if (m_histogram[i] > m_threshold) {
            flag = true;
            overflow = m_histogram[i] - m_threshold;
            overflow = overflow / HISTOGRAM_BIN_COUNT;
            m_histogram[i] = m_threshold;
            for (int k = 0; k < HISTOGRAM_BIN_COUNT; k++)
               m_histogram[k] += overflow;
         }
      }
      if (!flag)
         break;
   }
   count=0;
   for(int i=0;i < HISTOGRAM_BIN_COUNT;i++) {
      m_histogram_sum[i]=count;
      count += m_histogram[i];
   }
   if(m_spu)
      ztaFreeSharedMem(m_spu);
   m_spu=ztaBuildSpuBundle(1,SpuCallback,this,0,0);
   return m_spu;
}

int16_t GraphNodeEqualize::SpuCallback(int16_t input,void *pparm,uint32_t parm,uint32_t parm2) {
   static GraphNodeEqualize *instance=0;
   uint8_t v;
   int count,i;
   if(pparm)
      instance=(GraphNodeEqualize *)pparm;
   if(input < 0)
      v=0;
   else if(input > 255)
      v=255;
   else
      v=input;
   i=(((int)v+1+HISTOGRAM_BIN_SIZE-1)/HISTOGRAM_BIN_SIZE)-1;
   count = instance->m_histogram_sum[i];
   count += (instance->m_histogram[i]*((v&(HISTOGRAM_BIN_SIZE-1))+1))/HISTOGRAM_BIN_SIZE;
   count = (255*count)/(instance->m_nChannel*instance->m_w*instance->m_h);
   if (count > 255)   
      count = 255;
   return (int16_t)count;
}

void GraphNodeEqualize::Cleanup() {
   if(m_spu) {
      ztaFreeSharedMem(m_spu);
      m_spu=0;
   }
}

void GraphNodeEqualize::SetContrast(float _contrast) {
   m_contrast=_contrast;
   m_threshold = (int)(m_contrast*(((float)(m_nChannel*m_w*m_h))/HISTOGRAM_BIN_COUNT));
}

File: SW/apps/equalize/kernels/equalize.m
//----------------------------------------------------------------------------
// Copyright [2014] [Ztachip Technologies Inc]
//
// Author: Vuong Nguyen
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except IN compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to IN writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//------------------------------------------------------------------------------

#include "../../../base/ztalib.h"
#include "equalize.h"
#include "equalize.h"
#include "equalize.p.img"

// Perform histogram equalization

typedef struct {
   uint32_t input;
   uint32_t output;
   uint32_t output2;
   int nchannels;
   uint32_t equalize;
   int w;
   int h;
   struct {
      int channel;
      int extra_zero[2];
   } ws;
} Request;

extern void mycallback(int);

static void equalize(void *_p,int pid) {
   Request *req=(Request *)_p;
   int x,y;
   int step,step_x,step_y;
   int dx,dy;
   uint32_t input,output2,p;
   int i,j,count,extra_zero;
   int np,nt;
   int len;
   int src_w,src_h;

   input=req->input+req->w*req->h*req->ws.channel;
   output2=req->output2+req->w*req->h*req->ws.channel;
   np=NUM_PCORE;
   nt=NUM_THREAD_PER_CORE;
   step_x = kHistogramInSize*nt;
   step_y = np;
   dx=req->w;
   dy=req->h;
   src_w=req->w;
   src_h=req->h;

   dx=dx*dy;
   dy=1;
   src_w=src_w*src_h;
   src_h=1;
   step_x=step_x*step_y;
   step_y=1;

   // Split the job in the X direction
   if(pid==0) {
      dx=dx/2;
      src_w=src_w/2;
   } else {
      input=input+dx/2;
      output2=output2+dx/2;
      dx=dx-dx/2;
      src_w=src_w-src_w/2;
   }

   count=1000/kHistogramInSize;

   i=0;
   j=0;
   > EXE_LOCKSTEP(equalize::init,np,nt);
   ztaTaskYield();

   for(y=0;y < dy;y+=step_y) {
      for(x=0;x < dx;x += step_x) {
         >CAST(UINT8)PCORE(np)[:].equalize::in[0:kHistogramInSize*nt-1] <= CAST(UINT8)MEM(input,src_w)[x:x+step_x-1];

         if(req->equalize) {
            >CAST(UINT8)MEM(output2,src_w)[x:x+step_x-1] <= PROC(0) <= CAST(UINT8)SYNC PCORE(np)[:].equalize::in[0:kHistogramInSize*nt-1];
         } else {
            >CAST(UINT8)MEM(output2,src_w)[x:x+step_x-1] <= CAST(UINT8)SYNC PCORE(np)[:].equalize::in[0:kHistogramInSize*nt-1];
         }

         > EXE_LOCKSTEP(equalize::exe,np,nt);

         ztaTaskYield();

         i++;
         j+=(step_x*step_y);
         if(i >= count) {
            > EXE_LOCKSTEP(equalize::accumulate,np,nt);
            ztaTaskYield();
            i=0;
         }
      }
   }

   > EXE_LOCKSTEP(equalize::accumulate,np,nt);

   ztaTaskYield();

   > equalize::done.count <= INT16(nt);

   // Summarize results among the threads...
   > EXE_LOCKSTEP(equalize::done,np,kHistogramBinSize);
   ztaTaskYield();

   // Save results to SCRATCH...
   len=np*kHistogramBinSize*VECTOR_WIDTH;
   p=(pid==0)?0:len*2*2;
   >CAST(INT16)SCRATCH(p,len)[0:len-1] <= CAST(INT16)PCORE(np)[0:np-1].equalize::histogram_lo[0:kHistogramBinSize-1][:];
   p += len*2;
   >CAST(INT16)SCRATCH(p,len)[0:len-1] <= CAST(INT16)PCORE(np)[0:np-1].equalize::histogram_hi[0:kHistogramBinSize-1][:];

   req->ws.extra_zero[pid]=j-dx*dy;
}

// Aggregate the results from all the cores from both process 1 & 2

static void equalize_final(Request *req) {
   uint32_t p;
   int np,len,extra_zero;

   np=NUM_PCORE;
   len=np*kHistogramBinSize*VECTOR_WIDTH;
   p=0;
   >PCORE(np)[0].equalize::histogram_lo[0:kHistogramBinSize*np-1][:] <= CAST(INT16)SCRATCH(p,len)[0:len-1];
   p += len*2;
   >PCORE(np)[0].equalize::histogram_hi[0:kHistogramBinSize*np-1][:] <= CAST(INT16)SCRATCH(p,len)[0:len-1];
   p += len*2;
   >PCORE(np)[0].equalize::histogram_lo[kHistogramBinSize*np:kHistogramBinSize*2*np-1][:] <= CAST(INT16)SCRATCH(p,len)[0:len-1];
   p += len*2;
   >PCORE(np)[0].equalize::histogram_hi[kHistogramBinSize*np:kHistogramBinSize*2*np-1][:] <= CAST(INT16)SCRATCH(p,len)[0:len-1];

   if((2*np) > NUM_THREAD_PER_CORE)
      ztaAbort(0);

   > equalize::done.count <= INT16(2*np);
   > EXE_LOCKSTEP(equalize::done,1,kHistogramBinSize);

   extra_zero=req->ws.extra_zero[0]+req->ws.extra_zero[1];

   >CAST(INT16)PCORE(np)[0].equalize::histogram_hi[kHistogramBinSize][:] <= INT16(extra_zero/1000);
   >CAST(INT16)PCORE(np)[0].equalize::histogram_lo[kHistogramBinSize][:] <= INT16(extra_zero%1000);

   > EXE_LOCKSTEP(equalize::adjust_extra_zero,1,1);

   p=req->output+req->ws.channel*(kHistogramBinSize*VECTOR_WIDTH)*2*2;
   > CAST(INT16)MEM(p,kHistogramBinSize*VECTOR_WIDTH)[:] <= CAST(INT16)PCORE(np)[0].equalize::histogram_hi[0:kHistogramBinSize-1][:];
   p+=kHistogramBinSize*VECTOR_WIDTH*2;
   > CAST(INT16)MEM(p,kHistogramBinSize*VECTOR_WIDTH)[:] <= CAST(INT16)PCORE(np)[0].equalize::histogram_lo[0:kHistogramBinSize-1][:];
}

// Process request from host to do equalization/histogram

void kernel_equalize_exe(
   unsigned int _req_id,
   unsigned int _input,
   unsigned int _output,
   unsigned int _output2,
   int _nchannels,
   unsigned int _equalize,
   int _w,
   int _h
   ) 
{
   Request req;
   int i;
   
   ztaInitPcore(zta_pcore_img);
   ztaInitStream(_equalize);
      
   req.input=_input;
   req.output=_output;
   req.output2=_output2;
   req.nchannels=_nchannels;
   req.equalize=_equalize;
   req.w=_w;
   req.h=_h;
   
   for(i=0;i < req.nchannels;i++) {
      req.ws.channel=i;
      ztaDualHartExecute(equalize,&req);
      equalize_final(&req);
   }
   ztaJobDone(_req_id);
}

